cmd/ecr-credential-provider/main.go

```diff
@@ -153,19 +179,83 @@ func (e *ecrPlugin) getPrivateCredsData(ctx context.Context, imageHost string, i
 	}, nil
 }
 
-func (e *ecrPlugin) GetCredentials(ctx context.Context, image string, args []string) (*v1.CredentialProviderResponse, error) {
+func (e *ecrPlugin) buildCredentialsProvider(ctx context.Context, request *v1.CredentialProviderRequest, imageHost string) (aws.CredentialsProvider, error) {
+	var err error
+
+	arn, ok := request.ServiceAccountAnnotations["eks.amazonaws.com/ecr-role-arn"]
+	if !ok {
+		arn = os.Getenv("AWS_ECR_ROLE_ARN")
+	}
+	if arn == "" {
+		return nil, errors.New("no arn provided, cannot assume role using ServiceAccountToken")
+	}
+
+	if e.sts == nil {
+		region := ""
+		if imageHost != ecrPublicHost {
+			region = parseRegionFromECRPrivateHost(imageHost)
+		}
+		e.sts, err = stsProvider(ctx, region)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return aws.CredentialsProviderFunc(func(ctx context.Context) (aws.Credentials, error) {
+			assumeOutput, err := e.sts.AssumeRoleWithWebIdentity(ctx, &sts.AssumeRoleWithWebIdentityInput{
+				RoleArn:          aws.String(arn),
+				RoleSessionName:  aws.String("ecr-credential-provider"),
+				WebIdentityToken: aws.String(request.ServiceAccountToken),
+			})
+			if err != nil {
+				return aws.Credentials{}, fmt.Errorf("failed to assume role: %w", err)
+			}
+			return aws.Credentials{
+				AccessKeyID:     *assumeOutput.Credentials.AccessKeyId,
+				SecretAccessKey: *assumeOutput.Credentials.SecretAccessKey,
+				SessionToken:    *assumeOutput.Credentials.SessionToken,
+			}, nil
+		}),
+		nil
+}
+
+func (e *ecrPlugin) GetCredentials(ctx context.Context, request *v1.CredentialProviderRequest, args []string) (*v1.CredentialProviderResponse, error) {
 	var creds *credsData
 	var err error
 
-	imageHost, err := parseHostFromImageReference(image)
+	if request.Image == "" {
+		return nil, errors.New("image in plugin request was empty")
+	}
+
+	imageHost, err := parseHostFromImageReference(request.Image)
 	if err != nil {
 		return nil, err
 	}
 
+	var credentialsProvider aws.CredentialsProvider = nil
+	if request.ServiceAccountToken != "" {
+		credentialsProvider, err = e.buildCredentialsProvider(ctx, request, imageHost)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	if imageHost == ecrPublicHost {
-		creds, err = e.getPublicCredsData(ctx)
+		var optFns = []func(*ecrpublic.Options){}
+		if credentialsProvider != nil {
+			optFns = append(optFns, func(o *ecrpublic.Options) {
```

Comment:
argh, I didn't notice earlier that these opt funcs take different Options types. That's annoying :( I thought we could pass the same []opts to both ecr and ecrpublic